### PR TITLE
adding new pickers

### PR DESCRIPTION
**Pull request by patriciogonzalezvivo**
Solving some fallbacks on the color picker, adding new directional picker as a link only on the GLSL blocks... cleaning... re arragening

### CODE REVIEW

File: src/js/addons/pickers/ColorPicker.js [hunk 1]
```diff
@@ -42,7 +45,7 @@ export default class ColorPicker extends Picker {
             let barcursors = document.createElement('div');
             let leftcursor = document.createElement('div');
             let rightcursor = document.createElement('div');
-
+
```
Comment (louh) [inline src/js/addons/pickers/ColorPicker.js]: Trailing whitespace :cry:

File: src/js/addons/pickers/ColorPicker.js [hunk 2]
```diff
@@ -145,10 +164,18 @@ export default class ColorPicker extends Picker {
...
+            let height = this.disc.height / ratio;
+            this.dom.colorDisc.width = width * ratio;
+            this.dom.colorDisc.height = height* ratio;
+            diskContex.scale(ratio, ratio);
+
             drawDisk(
-                colorDisc.getContext('2d'),
-                [colorDisc.width / 2, colorDisc.height / 2],
-                [colorDisc.width / 2 - 1, colorDisc.height / 2 - 1],
+                diskContex,
```
Comment (louh) [inline src/js/addons/pickers/ColorPicker.js]: `diskContex` :arrow_right: `diskContext`

File: src/js/addons/pickers/ColorPicker.js [hunk 3]
```diff
@@ -145,10 +164,18 @@ export default class ColorPicker extends Picker {
...
+                [width/2, height/2],
+                [width/2 - 1, height/2 - 1],
```
Comment (louh) [inline src/js/addons/pickers/ColorPicker.js]: Please add spaces around operators, generally a good practice for readability, see, for example:
- http://airbnb.io/javascript/#18.4
- http://standardjs.com/rules.html

This should be enforced by the linter.

File: src/js/addons/pickers/ColorPicker.js [hunk 4]
```diff
@@ -159,18 +186,22 @@ export default class ColorPicker extends Picker {
...
+                [width/2, height/2],
                 '#303030',
-                2
+                2/ratio
             );
 
             // draw the luminanceBar bar
             let ctx = this.dom.luminanceBar.getContext('2d');
-            let gradient = ctx.createLinearGradient(0, 0, 0, 200);
+            this.dom.luminanceBar.width = this.barlum.width ;
```
Comment (louh) [inline src/js/addons/pickers/ColorPicker.js]: Extra space before `;`

File: src/js/addons/pickers/Matrix.js [hunk 5]
```diff
@@ -0,0 +1,100 @@
+'use strict'
+
+import Vector from './Vector';
+
+// TODO:
+// 		- generalize this for mat2, mat3 and mat4
+
+export default class Matrix {
+	constructor(m, type) {
```
Comment (louh) [inline src/js/addons/pickers/Matrix.js]: It seems there is mixed spaces/tabs in this file?

File: src/js/addons/pickers/Picker.js [hunk 6]
```diff
@@ -54,11 +51,18 @@ export default class Picker {
...
-        this.canvas.width = this.width;
-        this.canvas.height = this.height;
+        // this.canvas.width = this.width;
+        // this.canvas.height = this.width;
 
         this.el.appendChild(this.canvas);
         this.ctx = this.canvas.getContext('2d');
+
+        // Sets canvas to the proper dimensions and at the correct pixel density
+        // Thanks Lou https://github.com/louh/bart/blob/gh-pages/main.js#L19-L23
```
Comment (louh) [inline src/js/addons/pickers/Picker.js]: Thanks for the shoutout but this is pretty standard stuff :smiley: 
Not necessary to crowd the code with this!

File: src/js/addons/pickers/common.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
+'use strict';
+
+// Thanks Lou Huang 
+// https://github.com/louh/bart/blob/gh-pages/main.js#L327-L335
```
Comment (louh) [inline src/js/addons/pickers/common.js]: Same here -- and if you need to refer to a source, please find something more authoritative :)

File: src/js/addons/pickers/common.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+export function getDevicePixelRatio (ctx) {
+    let devicePixelRatio = window.devicePixelRatio || 1
```
Comment (louh) [inline src/js/addons/pickers/common.js]: We should enforce semicolons here

File: src/js/core/common.js [hunk 9]
```diff
@@ -143,19 +143,42 @@ export class StopWatch {
...
-        on(type, f) {
-            let listener = {};
-            listener[type] = f;
+        off (type, f) {
+            if (f) {
+                let listener = {};
+                listener[type] = f;
+                listeners.delete(listener);
+            }
+            else {
```
Comment (louh) [inline src/js/core/common.js]: Can you document what subscribeMixin is supposed to do? I should have asked earlier but never did...
We should also figure out what you're trying to accomplish and try to do it in a way that doesn't need as many nested loops / ifs

**Comment by patriciogonzalezvivo** [comment]
@louh Linted and checked

**Comment by patriciogonzalezvivo** [comment]
Once this merge, I will add the Vec3Picker with normalized values to lights that have directions, together with FloatPicker to simple numbers that don't need pressition

**Pull request merged**